clippy_lints/src/from_instead_of_into.rs

```diff
@@ -0,0 +1,82 @@
+use crate::utils::snippet;
+use crate::utils::span_lint_and_sugg;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir::*;
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_lint_pass, declare_tool_lint};
+use rustc_span::symbol::sym;
+
+declare_clippy_lint! {
+    /// **What it does:** Checking for using of From or TryFrom trait as a generic bound.
+    ///
+    /// **Why is this bad?** Into and TryInto are supersets of From and TryFrom. Due to
+    /// coherence rules, sometimes From and TryFrom are forbid to implemented but Into and
+    /// TryInto are not. So Into is a more generic bound than From, We should choose Into or
+    /// TryInto instead of From or TryFrom.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust
+    /// fn foo<T>(a: T) where u32: From<T> {}
+    /// fn bar<T>(a: T) where u32: TryFrom<T> {}
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// fn foo<T>(a: T) where T: Into<u32> {}
+    /// fn bar<T>(a: T) where T: TryInto<u32> {}
+    /// ```
+    pub FROM_INSTEAD_OF_INTO,
+    style,
+    "Into or TryInto trait is a better choice than From or TryFrom trait as a generic bound"
+}
+
+declare_lint_pass!(FromInsteadOfInto => [FROM_INSTEAD_OF_INTO]);
+
+impl LateLintPass<'tcx> for FromInsteadOfInto {
+    fn check_where_predicate(&mut self, cx: &LateContext<'tcx>, wp: &'tcx WherePredicate<'tcx>) {
+        match wp {
+            WherePredicate::BoundPredicate(wbp) => {
+                if_chain! {
+                    if let Some(tr_ref) = wbp.bounds[0].trait_ref();
```

Comment:
The ``From`` bound does not have to be the first bound in the predicate. So you could use `wbp.bounds.iter().find(|b| cond(b))` here instead.